priv/samples/basics/spi_rgb_leds.livemd

```diff
@@ -0,0 +1,371 @@
+                          ------------------ 
+</pre>
+
+<h5>If using an external power supply, ensure it shares a common Ground with your board.</h5>
+
+</details>
+
+You should be able to see your enabled SPI buses with the following:
+
+```elixir
+Circuits.SPI.bus_names()
+```
+
+Lets match on your bus.
+
+```elixir
+{:ok, ref} = Circuits.SPI.open("spidev0.0")
+```
+
+[return to top](#table-of-contents)
+
+## Formatting our SPI Serial Data Structure
+
+Now that we have our LED strip wired up to our device, let's send some data!
+
+The SK9822 and APA102 leverage the same series data structure. 
+A thorough write up on a unified protocol is available [here][SK9822-a-clone-of-the-apa102] 
+and this is the same logic leveraged by the popular Arduino [FastLED library][fastLED].
+
+We will need to send a few frames from the controller to the peripheral.
+
+###### Start Frame of 32 `zero` Bits.
+
+| 0000 0000 | 0000 0000 | 0000 0000 | 0000 0000 |
+| :-------: | :-------: | :-------: | :-------: |
+|   0x00    |    0x00   |    0x00   |    0x00   |
+|   8 Bits  |   8 Bits  |   8 Bits  |   8 Bits  |
+
+Elixir example: `<<0::32>>`.
+
+###### LED Frame 32 Bits.
+
+| Gray Scale | Brigtness |   Blue    |   Green   |    Red    |
+|:----------:|:---------:|:---------:|:---------:|:---------:|
+|   111      |   11111   | 1111 1111 | 1111 1111 | 1111 1111 |
+|   3 Bits   |   5 Bits  |   8 Bits  |   8 Bits  |   8 Bits  |
+
+Elixir example: `<<3::3, 1::5, 0xFF, 0xFF, 0xFF>>`.
+
+###### End Frame
+
+|   0   |
+| :---: |
+|  0x0  |
+| 1 Byte |
+
+Times (n/2), where n is the number of LEDs on the strip rounded to the next byte.
+
+Elixir example: `<<0::8>>`.
+
+[return to top](#table-of-contents)
+
+## Blinking a single LED
+
+First, we will turn on your first LED.
+
+Start frame is static, and can be reused.
+
+We will only have one LED frame, to blink the first LED on the strip.
+
+However, our end frame depends on the number of LEDs we have, 
+so lets calculate the number of bytes we will end with.
+
+The `Kernel.<>/2` operator concatinates our binaries.
+
+```elixir
+led_count = 1
+start_frame = << 0::32 >>
+# The end frame is 8 '0' bits * (leds - 1) / 16 (rounded up)
+# These are clock edges to ensure the LED's respond immediately and is borrowed from the FASTLed library.
+end_frame = 
+  Enum.map(1..round((led_count - 1) / 16), fn _ -> <<0::8>> end)
+  |> Enum.map_join(& &1)
+
+# Creating a List of frames with each LED on white
+led_frame = <<7::3, 1::5, 0xff, 0xff, 0xff>>
+
+# Transfer the frames over spi
+Circuits.SPI.transfer(ref, start_frame <> led_frame <> end_frame)
+```
+
+Your first LED should bet lit up white now.
+
+Lets blink it a few times, alternating the `led_frame` and an `off_frame`.
+
+```elixir
+blink_times = 10
+delay = 1000
+off_frame = <<7::3, 1::5, 0x00, 0x00, 0x00>>
+
+Enum.map(0..blink_times, fn x ->
+    case rem(x, 2) do
+        0 -> Circuits.SPI.transfer(ref, start_frame <> led_frame <> end_frame)
```

Comment:
What do you think of aliasing `Circuits.SPI` and just typing `SPI.transfer` everywhere? The `Circuits.SPI.transfer` is seeming verbose to me.

Reply:
Definitely aliasing!